firmware/controllers/serial/serial_rx.cpp

```diff
@@ -0,0 +1,44 @@
+/**
+ * @file	serial_rx.cpp
+ *
+ * This file handles auxilery serial communication.
+ *
+ * @date Mar 19, 2020
+ * @author Konstantin Smola, (c) 2020
+ */
+
+#include "globalaccess.h"
+
+#if EFI_AUX_SERIAL
+#include "engine.h"
+#include "serial.h"
+#include "serial_hw.h"
+#include "serial_sensor.h"
+
+#include "allsensors.h"
+#include "vehicle_speed.h"
+
+EXTERN_ENGINE;
+
+static LoggingWithStorage logger("AUX Serial RX");
+
+uint8_t ser_buffer[SERBUFFLEN] = {};
+uint16_t innovate_msg_len = SERBUFFLEN;
+innovate_serial_id_state_t innovate_serial_id_state = UNKNOWN;
+
+SerialRead::SerialRead()
+	: ThreadController("AUX Serial RX", NORMALPRIO) {
+}
+
+void SerialRead::ThreadTask() {
+	while (true) {
+		if (CONFIG(enableInnovateLC2)) {
+			len = innovate_msg_len;
+		}
+
+		if ((sdReadTimeout(AUX_SERIAL_PORT, ser_buffer, len, TIMEREAD)) > 0)
```

Comment:
what happens if we read a partial message? For example, what happens if we call read after `14.73` has been partially sent, so we rx `14.`, with 2 bytes still in the buffer?

Reply:
(these are binary formatted, not ascii, but you get the point: we could still do a partial read here)

Reply:
Good point, how about: `if ((sdReadTimeout(AUX_SERIAL_PORT, ser_buffer, len, TIMEREAD)) == len)`

Reply:
don't you want
```
while (bytesRead < target) {
    bytesRead += read(...);
}
```

Reply:
or even something more intelligent like reading single bytes until we find a header, then read exactly the correct number

Reply:
well, that's what we are doing, we are reading the whole buffer size, until we find the right size in IdentifyInnovateSerialMsg() and then len is set to the found msg size